### DISCUSSION TITLE
Labels missing in Output

### DISCUSSION THREAD

**Discussion by sthellmuth** [prometheus/client_python#900] [Q&A] [answered]
Hello,

i am trying to set both, labels and label values at time of instantiating a prometheus metric e.g. a Gauge.

I do this by passing a list of label names via the labelnames parameter and also a list of values by the _labelvalues parameter.

`self.prom_labels = ['Department','Line','Name','Site']`
`self.prom_labels_values = ['2','foo','my station','1']`

`self.prom_val = Gauge(
name=self.text,
            documentation=self.prom_doc,
            labelnames=self.prom_labels,
            _labelvalues=self.prom_labels_values,
        )
        `

`REGISTRY.register(self.prom_val))`

later on I set the value of my Gauge like this:
`self.prom_val.set(value=new_value)`

Here is the Output I can see under (localhost:8080/metrics)

OUTPUT:
\# HELP Test_Dint_01 Example doc1
\# TYPE Test_Dint_01 gauge
Test_Dint_01 0.0 **_...no labels here_**

However if I'm just setting the labelnames at the time of creating the instance of Gauge and passing the values at the time of setting the value it's working like expected and I can see them in the Output
Since the labelname labelvalues comes from a config file and i would like to be flexible about the number of labels this is not an an option for me.
...i also did not find a way to pass a list to the .labels method

`self.prom_labels = ['Department','Line','Name','Site']`
`self.prom_labels_values = ['2','foo','my station','1']`

`self.prom_val = Gauge(
            name=self.text, documentation=self.prom_doc, labelnames=self.prom_labels
        )`

`self.prom_val.labels(
                self.prom_labels_values[0],
                self.prom_labels_values[1],
                self.prom_labels_values[2],
                self.prom_labels_values[3],
            ).set(value=new_value)`
            
OUTPUT:
\# HELP Test_Dint_01 Example doc1
\# TYPE Test_Dint_01 gauge
Test_Dint_01{Department="2",Line="foo",Name="my station",Site="1"} 0.0

I found a way to solve my problem [here](https://stackoverflow.com/a/7745986)
So with a dict of labelnames --> labelvalues i can pass the labelnames at construction time and the labels values at the time of setting the Gauge value

`self.prom_labels_dict = {'Department':'2','Line':foo,'Name':'my station','Site':'1'}`
`self.prom_val = Gauge(
            name=self.text, documentation=self.prom_doc, labelnames=self.prom_labels
        )`
`self.prom_val.labels(**self.prom_labels_dict).set(value=new_value)`

where **self.prom_labels_dict automagically converts the dict to keyword arguments for the .labels method
took some time for me to figure it out so maybe this discussion helps somebody else 

Is there a better way to do such things?
Let me know if you need more Information

**Comment by csmarchbanks** [comment, marked as answer]
Hello, to answer a few of your questions!

> I do this by passing a list of label names via the labelnames parameter and also a list of values by the _labelvalues parameter.

You should not be using `_labelvalues`, it is for internal use only hence being started with an underscore which is the python convention for private variables. Always use `.labels()`.

> Since the labelname labelvalues comes from a config file and i would like to be flexible about the number of labels this is not an an option for me.

The best way to do this is to always use the same number of labels, but set the value of the label you do not need to `""`. Prometheus treats an empty string the same as if the label was missing.

> ...i also did not find a way to pass a list to the .labels method

It looks like you found the answer in the linked stack overflow, but just to keep all the info in one place: You should be able to pass them in with `*labelvalues`, so in your example:
```
self.prom_val.labels(*self.prom_labels_values).set(value=new_value)
```
The dict option you have is also fully supported.

I think that covered everything! Let me know if I missed anything/could make anything clearer.
